### PR TITLE
Feature/complete middleware saga

### PR DESCRIPTION
**Pull request by RobertWSaunders**
*Fixes ticket #25 *

### Changes proposed in this PR:
- updated custom structure for hacker middleware to digest, the previous structure would not work with sagas as it did not have a `type` property
- completed saga implementation for refresh token logic, if an api request is made and the reponds saying that the token expires, we will try refreshing once, if that succeeds we will make the original api request again with the new token. If the refreshing does not succeed it will simply log the user out. If an api request is made when tokens are being retrieved the request will be buffered until new tokens are found, at which point it will be dispatched. 

For testing purpose I used the GET `users`, so that is why there are actions for it although they are not super useful for what we need right now we will need them eventually so I thought I'd leave them in, also provides a good example of how to make a request that requires a token in our custom middleware.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -31,6 +31,7 @@
     "arrow-parens": ["error", "always"],
     "func-style": ["error", "declaration"],
     "react/jsx-wrap-multilines": "error",
-    "prefer-const": "error"
+    "prefer-const": "error",
+    "no-debugger": "error"
```
Comment (zackharley) [inline .eslintrc]: Why was this rule needed? Just curious, I've never actually used `debugger` before

File: package.json [hunk 2]
```diff
@@ -5,11 +5,11 @@
   "author": "Robert Saunders, Zack Harley, Cory Crowley & Avi Ihejirika",
   "scripts": {
     "server-dev": "cross-env NODE_ENV=development nodemon server",
-    "server-debug": "cross-env NODE_ENV=development nodemon --inspect server",
-    "start": "cross-env NODE_ENV=production npm run clean && npm run client-build && node server",
+    "server-debug": "cross-env NODE_ENV=development --inspect server",
```
Comment (zackharley) [inline package.json]: This is broken, missing `nodemon`

File: package.json [hunk 3]
```diff
@@ -5,11 +5,11 @@
...
+    "start": "cross-env NODE_ENV=production npm run clean && npm run build && node server",
     "client-dev": "cross-env webpack-dev-server",
-    "client-build": "cross-env NODE_ENV=production npm run clean && webpack -p",
-    "test": "cross-env eslint server/**/*.js client/**/*.js && nsp check --reporter checkstyle && ncu -e 2",
+    "build": "cross-env NODE_ENV=production npm run clean && webpack -p",
```
Comment (zackharley) [inline package.json]: I prefer the name-spaced approach you've taken with scripts thus far. If we can, I'd like to keep it as `client-build` or even `client-prod`

File: package.json [hunk 4]
```diff
@@ -26,9 +26,9 @@
     "jsonwebtoken": "^8.1.0",
     "lodash": "^4.17.4",
     "moment": "^2.19.1",
-    "mongoose": "^4.13.2",
-    "react": "^16.1.1",
-    "react-dom": "^16.1.1",
+    "mongoose": "^4.13.1",
```
Comment (zackharley) [inline package.json]: Outdated dependency

File: package.json [hunk 5]
```diff
@@ -26,9 +26,9 @@
...
+    "react": "^16.1.0",
```
Comment (zackharley) [inline package.json]: Outdated dependency

File: package.json [hunk 6]
```diff
@@ -26,9 +26,9 @@
...
+    "react-dom": "^16.1.0",
```
Comment (zackharley) [inline package.json]: Outdated dependency

File: .eslintrc (hunk 1 above)
Comment (RobertWSaunders) [reply .eslintrc]: Sometimes I `debugger` to step through the `ui`, so just added this rule to help me remember to remove them.

**Comment by zackharley** [comment]
@RobertWSaunders What operations did you perform to rebase this? Looks like a shit ton of commits

**Comment by RobertWSaunders** [comment]
@zackharley had to rebase twice on dev as PR was merged before this one, can we merge this one on, currently blocking.

**Pull request merged**